### PR TITLE
[Notification Template API] Event Handler Implementation

### PR DESCRIPTION
**Pull request by RushanNanayakkara**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Related Issues
- https://github.com/wso2/product-is/issues/21209

## Related PRs
- https://github.com/wso2-extensions/identity-governance/pull/832 [Dependent]
- https://github.com/wso2/identity-api-server/pull/629
- https://github.com/wso2/carbon-identity-framework/pull/5784

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -477,7 +477,7 @@
 
     <properties>
         <!--Identity Governance Version-->
-        <identity.governance.version>1.10.0</identity.governance.version>
+        <identity.governance.version>1.11.4-SNAPSHOT</identity.governance.version>
```
Comment (RushanNanayakkara) [inline pom.xml:480]: The snapshot versions need to be bumped after dependent PRs are merged. The snapshot values will be replaced with new release versions.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/internal/I18nMgtServiceComponent.java [hunk 2]
```diff
@@ -110,6 +112,20 @@ protected void activate(ComponentContext context) {
                 log.error("Error registering Email Template Mgt Service.");
             }
 
+            // Register Notification Template Mgt Service as an OSGi service.
+            NotificationTemplateManagerImpl notificationTemplateManager = new NotificationTemplateManagerImpl();
+            Hashtable<String, String> serviceProperties = new Hashtable<>();
+            serviceProperties.put("service.name", "NotificationTemplateManager");
```
Comment (AmshikaH) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/internal/I18nMgtServiceComponent.java:118]: Shall we move these to constants?

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/internal/I18nMgtServiceComponent.java [hunk 3]
```diff
@@ -110,6 +112,20 @@ protected void activate(ComponentContext context) {
...
+            ServiceRegistration notificationTemplateSR = bundleCtx.registerService(
+                    NotificationTemplateManager.class.getName(), notificationTemplateManager, serviceProperties);
+            if (notificationTemplateSR != null) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Notification Template Mgt Service registered.");
+                }
```
Comment (AmshikaH) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/internal/I18nMgtServiceComponent.java:124]:
```suggestion
                log.debug("Notification Template Mgt Service registered.");
```

As this is a simple debug statement with no concatenations or additional processing, the if statement is not necessary;`log.debug` will only log the statement if debug logs are enabled.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/internal/I18nMgtServiceComponent.java (hunk 2 above)
Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/internal/I18nMgtServiceComponent.java]: Addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/250/commits/b828356ab48d2c21dbdf01a3949e3ec86b98ce14

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/internal/I18nMgtServiceComponent.java (hunk 3 above)
Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/internal/I18nMgtServiceComponent.java]: Addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/250/commits/b828356ab48d2c21dbdf01a3949e3ec86b98ce14

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java [hunk 4]
```diff
@@ -0,0 +1,720 @@
...
+         } catch (OrganizationManagementException e) {
+             throw new NotificationTemplateManagerServerException(e.getMessage(), e);
+         }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void addNotificationTemplateType(String notificationChannel, String displayName, String tenantDomain)
```
Comment (RushanNanayakkara) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java:96]: truncate the displayName to remove any surrounding empty spaces

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java [hunk 5]
```diff
@@ -0,0 +1,720 @@
...
+/**
+ * Provides functionality to manage notification templates.
+ */
+public class NotificationTemplateManagerImpl implements NotificationTemplateManager {
+
+    private static final Log log = LogFactory.getLog(NotificationTemplateManagerImpl.class);
+
+    private final TemplatePersistenceManager userDefinedTemplatePersistenceManager;
+    private final TemplatePersistenceManager systemTemplatePersistenceManager;
+    private final TemplatePersistenceManager unifiedTemplatePersistenceManager;
```
Comment (RushanNanayakkara) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java:57]: Rename to a general name : eg : templatePersistenceManager

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java [hunk 6]
```diff
@@ -0,0 +1,720 @@
...
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void addNotificationTemplateType(String notificationChannel, String displayName, String tenantDomain)
+            throws NotificationTemplateManagerException {
+
+        validateDisplayNameOfTemplateType(displayName);
+        try {
```
Comment (RushanNanayakkara) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java:100]: At adding , validate to not include any leading or following empty spaces.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java [hunk 7]
```diff
@@ -0,0 +1,720 @@
...
+        } catch (IdentityApplicationManagementServerException e) {
+            String code = I18nEmailUtil.prependOperationScenarioToErrorCode(
+                    TemplateMgtConstants.ErrorMessages.ERROR_CODE_ERROR_RESOLVING_MAIN_APPLICATION.getCode(),
+                    TemplateMgtConstants.ErrorScenarios.NOTIFICATION_TEMPLATE_MANAGER);
+            String message = String.format(
+                    TemplateMgtConstants.ErrorMessages.ERROR_CODE_ERROR_RESOLVING_MAIN_APPLICATION.getMessage(),
+                    applicationUuid, tenantDomain);
+            throw new NotificationTemplateManagerException(code, message, e);
+        }
+        validateTemplateLocale(locale);
```
Comment (RushanNanayakkara) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java:282]: Move code duplicaitons to a shared place is possible.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java [hunk 8]
```diff
@@ -0,0 +1,720 @@
...
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<NotificationTemplate> getAllSystemNotificationTemplatesOfType(String notificationChannel,
+                                              String templateDisplayName) throws NotificationTemplateManagerException {
+
+        assertSystemTemplateTypeExists(templateDisplayName, notificationChannel);
```
Comment (RushanNanayakkara) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java:323]: Rename to a better function name.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java [hunk 9]
```diff
@@ -0,0 +1,720 @@
...
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void resetNotificationTemplateType(String notificationChannel, String templateType, String tenantDomain)
+            throws NotificationTemplateManagerException {
+        try {
```
Comment (RushanNanayakkara) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java:529]: add a empty line

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java [hunk 10]
```diff
@@ -0,0 +1,720 @@
...
+            if (OrganizationManagementUtil.isOrganization(tenantDomain)) {
+                // Return the root organization's email templates.
+                tenantDomain = getRootOrgTenantDomain(tenantDomain);
+            }
+            assertTemplateTypeExists(templateDisplayName, notificationChannel, tenantDomain);
+            List<NotificationTemplate> notificationTemplates =
+                    userDefinedTemplatePersistenceManager.listNotificationTemplates(templateDisplayName,
+                            notificationChannel, applicationUuid, tenantDomain);
+            if (notificationTemplates == null) {
```
Comment (RushanNanayakkara) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java:234]: if not returning a null remove this null check.

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java [hunk 11]
```diff
@@ -0,0 +1,720 @@
...
+    private String getDefaultNotificationLocale(String notificationChannel) {
+
+        if (NotificationChannels.SMS_CHANNEL.getChannelType().equals(notificationChannel)) {
+            return DEFAULT_SMS_NOTIFICATION_LOCALE;
+        } else {
+            return DEFAULT_EMAIL_NOTIFICATION_LOCALE;
+        }
+    }
+
+    private void assertSystemTemplateTypeExists(String templateType, String notificationChannel)
```
Comment (RushanNanayakkara) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java:691]: rename all assert function to better function names

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/constants/TemplateMgtConstants.java [hunk 12]
```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2024, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
+ *
```
Comment (RushanNanayakkara) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/constants/TemplateMgtConstants.java:3]: check the validity of the added licence headers

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/internal/I18nMgtServiceComponent.java [hunk 13]
```diff
@@ -71,6 +73,9 @@
 import javax.xml.stream.XMLStreamReader;
 
 import static org.wso2.carbon.email.mgt.constants.I18nMgtConstants.NOTIFICATION_TEMPLATES_LEGACY_TENANTS;
+import static org.wso2.carbon.email.mgt.constants.I18nMgtConstants.SERVICE_PROPERTY_KEY_SERVICE_NAME;
+import static org.wso2.carbon.email.mgt.constants.I18nMgtConstants.SERVICE_PROPERTY_VAL_EMAIL_TEMPLATE_MANAGER;
+import static org.wso2.carbon.email.mgt.constants.I18nMgtConstants.SERVICE_PROPERTY_VAL_NOTIFICATION_TEMPLATE_MANAGER;
 import static org.wso2.carbon.email.mgt.constants.I18nMgtConstants.SMS_PROVIDER;
 import static org.wso2.carbon.email.mgt.constants.I18nMgtConstants.SMS_PROVIDER_POST_BODY_TEMPLATES_DIR_PATH;
 import static org.wso2.carbon.email.mgt.constants.I18nMgtConstants.TEMPLATE_BODY;
```
Comment (RushanNanayakkara) [inline components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/internal/I18nMgtServiceComponent.java:81]: Add an issue to rename this class to a more appropirate name;

Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/internal/I18nMgtServiceComponent.java:81]: tracked through https://github.com/wso2/product-is/issues/21454

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java (hunk 4 above)
Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java:96]: addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/250/commits/4a4dae818d530081e8a91a33d6d70da13c375c9c

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java (hunk 5 above)
Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java]: addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/250/commits/4a4dae818d530081e8a91a33d6d70da13c375c9c

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java (hunk 6 above)
Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java:103]: addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/250/commits/4a4dae818d530081e8a91a33d6d70da13c375c9c

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java (hunk 7 above)
Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java:280]: addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/250/commits/4a4dae818d530081e8a91a33d6d70da13c375c9c

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java (hunk 8 above)
Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java]: addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/250/commits/4a4dae818d530081e8a91a33d6d70da13c375c9c

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java (hunk 9 above)
Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java:526]: addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/250/commits/4a4dae818d530081e8a91a33d6d70da13c375c9c

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java (hunk 10 above)
Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java]: addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/250/commits/4a4dae818d530081e8a91a33d6d70da13c375c9c

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java (hunk 11 above)
Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java]: addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/250/commits/4a4dae818d530081e8a91a33d6d70da13c375c9c

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java [hunk 14]
```diff
@@ -0,0 +1,720 @@
...
+
+        if (StringUtils.isBlank(locale)) {
+            String errorCode =
+                    I18nEmailUtil.prependOperationScenarioToErrorCode(
+                            TemplateMgtConstants.ErrorMessages.ERROR_CODE_EMPTY_LOCALE.getCode(),
+                            TemplateMgtConstants.ErrorScenarios.NOTIFICATION_TEMPLATE_MANAGER);
+            throw new NotificationTemplateManagerClientException(errorCode,
+                    TemplateMgtConstants.ErrorMessages.ERROR_CODE_EMPTY_LOCALE.getMessage());
+        }
+        final String LOCAL_REGEX = "^[a-z]{2}_[A-Z]{2}$";
```
[the remark this answers is not in the document]
Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java]: addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/250/commits/4a4dae818d530081e8a91a33d6d70da13c375c9c

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java [hunk 15]
```diff
@@ -0,0 +1,720 @@
...
+        if (!locale.matches(LOCAL_REGEX)) {
```
[the remark this answers is not in the document]
Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java]: addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/250/commits/4a4dae818d530081e8a91a33d6d70da13c375c9c

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/constants/TemplateMgtConstants.java [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+    public static class ErrorScenarios {
+
+        // NTM - NOTIFICATION TEMPLATE MANAGER
+        public static final String NOTIFICATION_TEMPLATE_MANAGER = "NTM";
+    }
+
+    /**
+     * Class which contains the error codes for template management.
+     */
+    public static class ErrorCodes {
```
[the remark this answers is not in the document]
Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/constants/TemplateMgtConstants.java:40]: addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/250/commits/4a4dae818d530081e8a91a33d6d70da13c375c9c

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/RegistryBasedTemplateManager.java [hunk 17]
```diff
@@ -297,6 +297,23 @@ public void deleteNotificationTemplates(String displayName, String notificationC
         }
     }
 
+    @Override
+    public void deleteAllNotificationTemplates(String displayName, String notificationChannel, String tenantDomain)
+            throws NotificationTemplateManagerServerException {
+
+        String templateType = I18nEmailUtil.getNormalizedName(displayName);
```
[the remark this answers is not in the document]
Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/store/RegistryBasedTemplateManager.java]: addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/250/commits/4a4dae818d530081e8a91a33d6d70da13c375c9c

File: components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java (hunk 7 above)
Comment (RushanNanayakkara) [reply components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java:280]: This is not complete code duplication and only the logic. Hence will be done seperately after having an internal discussion.

**Review by Avarjana — APPROVED** [review]

**Pull request merged**